src/Mono.Android/Android.Runtime/AndroidRuntime.cs

```diff
@@ -248,9 +248,80 @@ static Delegate CreateDynamicCallback (MethodInfo method)
 			return (Delegate)dynamic_callback_gen.Invoke (null, new object [] { method });
 		}
 
+		static List<JniNativeMethodRegistration> sharedRegistrations = new List<JniNativeMethodRegistration> ();
+
+		static bool FastRegisterNativeMembers (JniType nativeClass, Type type, string methods)
+		{
+			if (!MagicRegistrationMap.Filled)
+				return false;
+
+			bool lockTaken = false;
+			bool rv = false;
+
+			try {
+				Monitor.TryEnter (sharedRegistrations, ref lockTaken);
+				List<JniNativeMethodRegistration> registrations;
+				if (lockTaken) {
+					sharedRegistrations.Clear ();
+					registrations = sharedRegistrations;
+				} else {
+					registrations = new List<JniNativeMethodRegistration> ();
+				}
+				JniNativeMethodRegistrationArguments arguments = new JniNativeMethodRegistrationArguments (registrations, methods);
+				rv = MagicRegistrationMap.CallRegisterMethod (arguments, type.FullName);
+
+				if (registrations.Count > 0)
+					nativeClass.RegisterNativeMethods (registrations.ToArray ());
+			} finally {
+				if (lockTaken) {
+					Monitor.Exit (sharedRegistrations);
+				}
+			}
+
+			return rv;
+		}
+
+		class MagicRegistrationMap {
+			static Dictionary<string, int> typesMap;
+
+			static void Prefill ()
+			{
+				// fill code added by the linker
+			}
+
+			static MagicRegistrationMap ()
+			{
+				typesMap = new Dictionary<string, int> ();
```

Comment:
...which raises an observation: This line shouldn't be here.

Instead, `Prefill()` should do the `new Dictionary<string, int>()` execution, as `Prefill()` -- or the code which generates `Prefill()` -- *knows how big the dict should be*, and thus can *instead* generate `new Dictionary<string, int>(Count)`.